src/devices/src/virtio/file_traits.rs

```diff
@@ -460,3 +465,54 @@ impl FileReadWriteAtVolatile for DiskProperties {
         Ok(full_length)
     }
 }
+
+#[cfg(feature = "blk")]
+fn create_iovec<'a>(
+    disk: &DiskProperties,
+    bufs: &'a [VolatileSlice],
+    offset: u64,
+    head_buffer: &'a mut Option<IoBuffer>,
+) -> Result<IoVector<'a>> {
+    match disk.image_format() {
+        disk::ImageType::Raw => {
+            let (iovec, _guard) = IoVector::from_volatile_slice(bufs);
+            guard_header_bytes(disk.file(), iovec, offset, head_buffer)
+        }
+        _ => Ok(IoVector::from_volatile_slice(bufs).0),
+    }
+}
+
+#[cfg(feature = "blk")]
+fn guard_header_bytes<'a>(
+    image: &imago::SyncFormatAccess<imago::file::File>,
+    iovec: IoVector<'a>,
+    offset: u64,
+    head_buffer: &'a mut Option<IoBuffer>,
+) -> Result<IoVector<'a>> {
+    if (1..4).contains(&offset) || (offset + iovec.len() < 4) {
+        // Make sure the guest writes to the first four bytes as a whole. This is done to prevent a
+        // Raw disk image byte-by-byte, and in any order, writing the header of a formatted disk
+        // image into the first sector of the disk image.
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            "partial write to first four bytes of disk image",
+        ));
+    }
```

Comment:
If this is impossible, and the mitigation below is also irrelevant, then it seems like I'll just remove 46f797e0ed4b0d16308ac0810b3d9802a672faac altogether

Reply:
Dropped the last 2 commits. Should hopefully be up to date now.

Reply:
(By the way, by “impossible” I didn’t mean it’d be impossible to prevent byte-by-byte writes by the guest. Quite the opposite, I meant that it’s impossible for the guest to do byte-by-byte writes, so it wouldn’t be necessary for us to detect them.)